torch/fx/experimental/symbolic_shapes.py

```diff
@@ -170,6 +170,41 @@ def is_concrete_int(a: Union[int, SymInt]):
 def tensor_has_hints(t):
     return all(has_hint(s) for s in t.size())
 
+def free_symbols(val: Union[SymInt, torch.Tensor]) -> Set[sympy.Symbol]:
+    if isinstance(val, (SymInt, SymFloat)):
+        return val.node.expr.free_symbols
+    elif isinstance(val, (int, float, bool)):
+        return set()
+    elif isinstance(val, torch.Tensor):
+        r = set()
+        for s in val.size():
+            r |= free_symbols(s)
+        for s in val.stride():
+            r |= free_symbols(s)
+        r |= free_symbols(val.storage_offset())
+        return r
+    else:
+        raise AssertionError(f"cannot compute free_symbols of {val}")
+
+# WARNING: Don't use this on Dynamo produced graphs, they don't have meta
+# setup!
+def is_symbol_binding_fx_node(node) -> Optional[sympy.Symbol]:
+    if (
+        node.op == "placeholder" and
+        "val" in node.meta and
+        isinstance(node.meta["val"], torch.SymInt) and
+        isinstance(node.meta["val"].node.expr, sympy.Symbol)
```

Comment:
Checking my understanding: is it possible to have a SymInt-bound placeholder node, who's `node.meta["val"].node.expr` is something other than ` sympy.Symbol`? If so, when? (and if not maybe we should assert against it).

Reply:
Yes, is possible. While it is true that we only ever add placeholder nodes for fresh symbols, a symbol can get refined as we do more operations, and then it turns out that it's a composite. For example:

```
def f(s0, s1):
    assert s0 == s1 * 2
```

The s0 will become s1 * 2 and no longer a sympy.Symbol.

Reply:
Ah, right. Thanks for the example!